relentless/potential/pair.py

```diff
@@ -156,6 +157,213 @@ def _derivative(self, param, r, epsilon, sigma, **params):
+
+        """
+        # check that r and u have the right shape
+        if len(r) != self.num_knots:
+            raise ValueError('r must have the same length as the number of knots')
+        if len(u) != self.num_knots:
+            raise ValueError('u must have the same length as the number of knots')
+
+        # convert to r,knot form given the mode
+        rs = np.asarray(r, dtype=np.float64)
+        ks = np.asarray(u, dtype=np.float64)
+        if self.mode == 'diff':
+            # difference is next knot minus my knot, with last knot fixed at its current value
+            ks[:-1] -= ks[1:]
+
+        # make all knots variables, but hold all r and the last knot const
+        for i in range(self.num_knots):
+            self.coeff[pair]['r-{}'.format(i)] = core.Variable(rs[i],const=True)
+            self.coeff[pair]['knot-{}'.format(i)] = core.Variable(ks[i],const=(i==self.num_knots-1))
+
+    def _energy(self, pair, r, **params):
+        """Evaluates the spline potential energy.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        r : array_like
+            The value or values of r at which to evaluate the energy.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the energy evaluated at r.
+
+        """
+        r,u,s = self._zeros(r)
+        params = self.coeff[pair]
+        u = self._interpolate(params)(r)
+        if s:
+            u = u.item()
+        return u
+
+    def _force(self, pair, r, **params):
+        """Evaluates the spline force.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        r : array_like
+            The value or values of r at which to evaluate the force.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the force evaluated at r.
+
+        """
+        r,f,s = self._zeros(r)
+        params = self.coeff[pair]
+        f = -self._interpolate(params).derivative(r, 1)
+        if s:
+            f = f.item()
+        return f
+
+    def _derivative(self, pair, param, r, **params):
+        """Evaluates the spline parameter derivative using finite differencing.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        param : str
+            The parameter with respect to which to take the derivative (the knot coefficient).
+        r : array_like
+            The value or values of r at which to evaluate the derivative.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the derivative evaluated at r.
+
+        """
+        r,d,s = self._zeros(r)
+        h = 0.001
+        params = self.coeff[pair]
```

Comment:
You already have all of the knot params in `**params`, so you shouldn't be accessing the coefficients directly.